src/updater.h

```diff
@@ -135,13 +136,85 @@ class MetadataUpdater {
   MetadataStore* store_;
 };
 
+// Abstract class for a periodic timer.
+class Timer {
+ public:
+  // Returns the period of the timer (in seconds).
+  virtual std::chrono::seconds Period() const = 0;
+
+  // Initializes the timer.
+  virtual void Init() = 0;
+
+  // Waits for one period to pass.  Returns false if the timer was
+  // canceled while waiting.
+  virtual bool Wait() = 0;
+
+  // Cancels the timer.
+  virtual void Cancel() = 0;
+};
+
+// Implementation of a periodic timer parameterized over a clock type.
+template<typename Clock>
+class TimerImpl : public Timer {
+ public:
+  TimerImpl(int period_s, bool verbose, const std::string& name)
+      : period_(period_s), verbose_(verbose), name_(name) {}
+  std::chrono::seconds Period() const { return period_; }
+  void Init() {
+    timer_.lock();
+    if (verbose_) {
+      LOG(INFO) << "Locked timer for " << name_;
+    }
+  }
+  bool Wait() {
+    // An unlocked timer means we should stop updating.
+    if (verbose_) {
+      LOG(INFO) << "Trying to unlock the timer for " << name_;
+    }
+    auto start = Clock::now();
+    auto wakeup = start + period_;
+    bool done = true;
+    while (done && !timer_.try_lock_until(wakeup)) {
```

Comment:
What do you think about inverting the meaning of done as well?

```
bool valid_wakeup = false;

while (!valid_wakeup && !timer_.try_lock_until(wakeup)) {
```

Reply:
Done.